### ISSUE TITLE
Kamal and ssh-auth inside docker

### ISSUE THREAD

**Issue by vldmr-k** [basecamp/kamal#468]
Hi guys
I'm try to using kamal inside docker via alias

In my `~/.profile` I have next string
```
alias kamal="docker run -it --rm -v '${PWD}:/workdir' -v '${SSH_AUTH_SOCK}:/ssh-agent' -v /var/run/docker.sock:/var/run/docker.sock -e 'SSH_AUTH_SOCK=/ssh-agent' ghcr.io/basecamp/kamal:latest"
```

When I try to authenticate to remove host, I got error
```
kamal app details
  INFO [da4b2410] Running docker ps --filter label=service=pmpro.com.ua --filter label=role=web on 49.13.*.*
  ERROR (Net::SSH::AuthenticationFailed): Exception while executing on host 49.13.*.*: Authentication failed for user root@49.13.*.*
```

I checked, whick keys loaded to ssh-agent on my host machine, and I see that my ssh key already loaded
```
ssh-add -l
256 SHA256:6wCa5a4iwr4ve8pV7+K8nopkRXnPTnmEEwSj5Gji3+4  (ED25519)
```

And when I try to connect to the service directly from host machine to remove machine, everythink working well
```
ssh -T root@49.13.*.*
Linux h6srv 6.1.0-11-arm64 #1 SMP Debian 6.1.38-4 (2023-08-08) aarch64
```

Could you help me please, explain what the problem? And why kamal inside docker don't see ssh keys?
Thank you

**Comment by n4cr** [comment]
Having the same issue

**Comment by vldmr-k** [comment]
Any help?

**Comment by djmb** [comment]
Hi @vldmr-k, you can set the log level on the SSH connection as described here - https://kamal-deploy.org/docs/configuration#using-a-different-ssh-log-level. That might give you an idea of what is going wrong.

**Comment by clintmiller** [comment]
@vldmr-k I'm currently using Kamal in docker as well. On a Mac, w/ Docker for Mac, you've got to pass some the SSH agent socket a little bit different way, w/ te path `/run/host-services/ssh-auth.sock`. This is mentioned in their [documentation here](https://docs.docker.com/desktop/networking/). I also set an env var: `SSH_AUTH_SOCK=/run/host-services/ssh-auth.sock`.

**Comment by vldmr-k** [comment]
> @vldmr-k I'm currently using Kamal in docker as well. On a Mac, w/ Docker for Mac, you've got to pass some the SSH agent socket a little bit different way, w/ te path `/run/host-services/ssh-auth.sock`. This is mentioned in their [documentation here](https://docs.docker.com/desktop/networking/). I also set an env var: `SSH_AUTH_SOCK=/run/host-services/ssh-auth.sock`.

Yeah, thanks. debbugin of ssh connection helps me well.

Now, I have another problem.

```
permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Post "http://%2Fvar%2Frun%2Fdocker.sock/v1.24/auth": dial unix /var/run/docker.sock: connect: permission denied
```

I worry about this message
```
Post "http://%2Fvar%2Frun%2Fdocker.sock/v1.24/auth"
```
It's looks like kamal try to make http request to socket, is it legal ?

**Comment by n4cr** [comment]
I solved it by updating docker

**Comment by bst27** [comment]
I had a similiar issue on **Ubuntu 22.04**:

As described in the [docs](https://kamal-deploy.org/docs/installation) I created an alias for Kamal:
```
alias kamal="docker run -it --rm -v '${PWD}:/workdir' -v '/run/host-services/ssh-auth.sock:/run/host-services/ssh-auth.sock' -e SSH_AUTH_SOCK='/run/host-services/ssh-auth.sock' -v /var/run/docker.sock:/var/run/docker.sock ghcr.io/basecamp/kamal:latest"
```

Running `kamal setup` failed with the following error message:
```
kamal setup
  INFO [2e7e5689] Running /usr/bin/env mkdir -p .kamal on 123.123.123.123
  Finished all in 0.2 seconds
  ERROR (Net::SSH::AuthenticationFailed): Exception while executing on host 123.123.123.123: Authentication failed for user root@123.123.123.123
```

I checked if I was able to connect directly without Kamal and this looked fine:
```
ssh root@123.123.123.123
```

**TLDR:**
I was able to solve the authentication problem by modifying the alias command slightly:
```
alias kamal="docker run -it --rm -v '${PWD}:/workdir' -v '$SSH_AUTH_SOCK:/run/host-services/ssh-auth.sock' -e SSH_AUTH_SOCK='/run/host-services/ssh-auth.sock' -v /var/run/docker.sock:/var/run/docker.sock ghcr.io/basecamp/kamal:latest"
```

**Comment by sbounmy** [comment]
With 
`alias kamal="docker run -it --rm -v '${PWD}:/workdir' -v '/run/host-services/ssh-auth.sock:/run/host-services/ssh-auth.sock' -e SSH_AUTH_SOCK='/run/host-services/ssh-auth.sock' -v /var/run/docker.sock:/var/run/docker.sock ghcr.io/basecamp/kamal:latest"`

I tried
- enabling `eval "$(ssh-agent)"` because I had `Error connecting to agent: No such file or directory...`
- `ssh root@123.123.123.123` it works well
- ```ssh-add -l``` returns the correct key

@vldmr-k how did you solve it ? It keep prompting for password

`kamal setup`
```log
D, [2023-10-20T08:38:46.668967 #1] DEBUG -- net.ssh.transport.session[618]: establishing connection to 123.123.123.123:22
D, [2023-10-20T08:38:46.702759 #1] DEBUG -- net.ssh.transport.session[618]: connection established
I, [2023-10-20T08:38:46.702808 #1]  INFO -- net.ssh.transport.server_version[62c]: negotiating protocol version
D, [2023-10-20T08:38:46.702824 #1] DEBUG -- net.ssh.transport.server_version[62c]: local is `SSH-2.0-Ruby/Net::SSH_7.2.0 aarch64-linux-musl'
D, [2023-10-20T08:38:46.742317 #1] DEBUG -- net.ssh.transport.server_version[62c]: remote is `SSH-2.0-OpenSSH_8.9p1 Ubuntu-3ubuntu0.3'
I, [2023-10-20T08:38:46.742511 #1]  INFO -- net.ssh.transport.algorithms[640]: sending KEXINIT
D, [2023-10-20T08:38:46.742624 #1] DEBUG -- socket[654]: queueing packet nr 0 type 20 len 852
D, [2023-10-20T08:38:46.742674 #1] DEBUG -- socket[654]: sent 856 bytes
D, [2023-10-20T08:38:46.773619 #1] DEBUG -- socket[654]: read 1080 bytes
D, [2023-10-20T08:38:46.773690 #1] DEBUG -- socket[654]: received packet nr 0 type 20 len 1076
I, [2023-10-20T08:38:46.773718 #1]  INFO -- net.ssh.transport.algorithms[640]: got KEXINIT from server
I, [2023-10-20T08:38:46.773765 #1]  INFO -- net.ssh.transport.algorithms[640]: negotiating algorithms
D, [2023-10-20T08:38:46.773809 #1] DEBUG -- net.ssh.transport.algorithms[640]: negotiated:
* kex: ecdh-sha2-nistp521
* host_key: ssh-ed25519
* encryption_server: aes256-ctr
* encryption_client: aes256-ctr
* hmac_client: hmac-sha2-512-etm@openssh.com
* hmac_server: hmac-sha2-512-etm@openssh.com
* compression_client: none
* compression_server: none
* language_client: 
* language_server: 
D, [2023-10-20T08:38:46.773822 #1] DEBUG -- net.ssh.transport.algorithms[640]: exchanging keys
D, [2023-10-20T08:38:46.775797 #1] DEBUG -- socket[654]: queueing packet nr 1 type 30 len 148
D, [2023-10-20T08:38:46.775843 #1] DEBUG -- socket[654]: sent 152 bytes
D, [2023-10-20T08:38:46.811572 #1] DEBUG -- socket[654]: read 312 bytes
D, [2023-10-20T08:38:46.811616 #1] DEBUG -- socket[654]: received packet nr 1 type 31 len 292
D, [2023-10-20T08:38:46.815611 #1] DEBUG -- socket[654]: queueing packet nr 2 type 21 len 20
D, [2023-10-20T08:38:46.815652 #1] DEBUG -- socket[654]: sent 24 bytes
D, [2023-10-20T08:38:46.815674 #1] DEBUG -- socket[654]: received packet nr 2 type 21 len 12
D, [2023-10-20T08:38:46.815899 #1] DEBUG -- net.ssh.authentication.session[668]: beginning authentication of `root'
D, [2023-10-20T08:38:46.815928 #1] DEBUG -- socket[654]: using encrypt-then-mac
D, [2023-10-20T08:38:46.816019 #1] DEBUG -- socket[654]: queueing packet nr 3 type 5 len 32
D, [2023-10-20T08:38:46.816048 #1] DEBUG -- socket[654]: sent 100 bytes
D, [2023-10-20T08:38:46.848572 #1] DEBUG -- socket[654]: read 100 bytes
D, [2023-10-20T08:38:46.848705 #1] DEBUG -- socket[654]: received packet nr 3 type 6 len 32
D, [2023-10-20T08:38:46.848760 #1] DEBUG -- net.ssh.authentication.session[668]: trying none
D, [2023-10-20T08:38:46.848807 #1] DEBUG -- socket[654]: using encrypt-then-mac
D, [2023-10-20T08:38:46.848842 #1] DEBUG -- socket[654]: queueing packet nr 4 type 50 len 48
D, [2023-10-20T08:38:46.848963 #1] DEBUG -- socket[654]: sent 116 bytes
D, [2023-10-20T08:38:46.888654 #1] DEBUG -- socket[654]: read 100 bytes
D, [2023-10-20T08:38:46.888736 #1] DEBUG -- socket[654]: received packet nr 4 type 51 len 32
D, [2023-10-20T08:38:46.888765 #1] DEBUG -- net.ssh.authentication.session[668]: allowed methods: publickey,password
D, [2023-10-20T08:38:46.888801 #1] DEBUG -- net.ssh.authentication.methods.none[67c]: none failed
D, [2023-10-20T08:38:46.888821 #1] DEBUG -- net.ssh.authentication.session[668]: trying publickey
D, [2023-10-20T08:38:46.888885 #1] DEBUG -- net.ssh.authentication.agent[690]: connecting to ssh-agent
D, [2023-10-20T08:38:46.888946 #1] DEBUG -- net.ssh.authentication.agent[690]: sending agent request 1 len 50
D, [2023-10-20T08:38:46.889522 #1] DEBUG -- net.ssh.authentication.agent[690]: received agent packet 5 len 1
D, [2023-10-20T08:38:46.889559 #1] DEBUG -- net.ssh.authentication.agent[690]: sending agent request 11 len 0
D, [2023-10-20T08:38:46.889835 #1] DEBUG -- net.ssh.authentication.agent[690]: received agent packet 12 len 5
D, [2023-10-20T08:38:46.889868 #1] DEBUG -- net.ssh.authentication.session[668]: trying password
root@123.123.123.123's password: 
```

**Comment by sbounmy** [comment]
Nevermind, figured out.

1.  just reopened a new terminal
2. `ssh-add ~/.ssh/id_ed25519 (or id_rsa)`
3. `ssh-add -l` (verify it does show up)
4. `kamal setup` works

In my ~/.aliases
`alias kamal="docker run -it --rm -v '${PWD}:/workdir' -v '/run/host-services/ssh-auth.sock:/run/host-services/ssh-auth.sock' -e SSH_AUTH_SOCK='/run/host-services/ssh-auth.sock' -v /var/run/docker.sock:/var/run/docker.sock ghcr.io/basecamp/kamal:latest"`
